### DISCUSSION TITLE
Modular / recursive configuration with multiple pytest.ini in subfolders

### DISCUSSION THREAD

**Discussion by waszil** [pytest-dev/pytest#7732] [General] [answered]
Hi, i have been searching around about this, but could not find any info. My question is, is it possible somehow to use multiple pytest.ini files in subfolders? What I would like to achieve is that I have a package with several subfolders/subpackages that may contain python files that may contain test functions. These files may not have the `test_*.py` naming convention, so in these subfolders I create a pytest.ini file with a `python_files = my_file_pattern*.py` entry. Now, if I invoke pytest on the root level folder of my package, these pytest.ini files in the subfolders are not considered as I can tell. Is there any way to achieve something like this? Also, I could use this to specify different `filterwarnings` entries in different subfolders.
Thanks for any hints.

**Comment by graingert** [comment]
You can use https://docs.pytest.org/en/stable/reference.html#globalvar-collect_ignore_glob in a `conftest.py`

**Comment by waszil** [comment]
How would that help? I need the opposite, I don't want to ignore some files, I would like to additionally collect some. Meanwhile I have played around with `addinivalue_line('python_files', ...` in the `def pytest_configure` hook in conftest.py, but no luck.
I also have tried the `pytest_collection` hook, but it is only called if the conftest.py is in the root folder, not if it is in a subfolder.

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Multiple ini files are not supported and at the moment there are no plans for this, as currently there is no reasonable and time effective way to implement this as a working solution

Pytest itself has a hack in the root ini file to support this

*edit* see line 16 of our pyproject.toml (can't make a proper link on mobile)

**Comment by waszil** [comment]
Thanks. So I guess I would be better off by reorganizing the project and extracting all tests to a separate tests folder. However I still see this as a possible feature, because in our case it is a quite big codebase, and in this case it would be better to have the tests next to the actual modules, or even inside them.

**Comment by kbakk** [comment]
I suppose you could alternatively write a plugin to achieve this (not very familiar with this myself, but seems doable). 

https://hackebrot.github.io/pytest-tricks/customize_class_collection/

**Reply by waszil** [reply]
Thanks for the idea, indeed, I could do it with this. Here is my solution in a `conftest.py` in my package root:

```python
from pathlib import Path
from pytest import Module

package_loc = Path(__file__).parent

additional_modules = [
    package_loc / "subfolder" / "my_additional_file.py",
]


def pytest_collect_file(path, parent):
    if Path(path) in additional_modules:
        return Module.from_parent(fspath=path, parent=parent)
    else:
        return None

```

**Comment by iamtheef** [comment]
@waszil How did you solve this?
@RonnyPfannschmidt as of 2024 any new solution to this?

**Reply by waszil** [reply]
at the end i had no time for this and just reorganized all tests out into a separate folder.

**Comment by RonnyPfannschmidt** [comment]
closing this discussion as config merges are currently not supported and there are no plans to directly include such a feature and there also is no proposal on how to do this safe/sane

**Comment by mattculler** [comment]
For anyone else who got here from google looking for how to merge pytest configs: the Sept 9 2020 reference to "check the pytest pyproject.toml" points at a config that looks like this:

```toml
[tool.pytest.ini_options]
python_files = ["test_*.py", "*_test.py", "testing/*/*.py"]
```

Which sadly is a hack addressing a different part of OP's question.  I haven't found a good way to have multiple pytest config files.
